README
NAME

    Blosxom::Header - Missing interface to modify HTTP headers

SYNOPSIS

        use Blosxom::Header;

        my $header = Blosxom::Header->new();
  
        $header->set(
            type          => 'text/html;',
            status        => '304',
            cache_control => 'must-revalidate',
        );
        my $value = $header->get('status');           # 304 Not Modified
        my $bool  = $header->exists('cache_control'); # 1
        my @keys  = $header->keys();                  # ('type', 'status', 'cache_control')

        $header->remove('cache_control');
        @keys = $header->keys(); # ('type', 'status')

        $header->remove_all();
        @keys = $header->keys(); # ()

DESCRIPTION

    Blosxom, a weblog application, exports a global variable $header
    which is a hash reference. This application passes $header CGI::header()
    to generate HTTP headers.

    When plugin writers modify HTTP headers, they must write as follows:

        package foo;
        $blosxom::header->{'-status'} = '304 Not Modified';

    It's obviously bad practice. Blosxom misses the interface to modify
    HTTP headers.  

    This module allows you to modify them in an object-oriented way.
    If loaded, you might write as follows:

        my $header = Blosxom::Header->new();
        $header->set('status' => '304');

INSTALLATION

        % perl Makefile.PL
        % make
        % make test

DEPENDENCIES

    *   Blosxom 2.1.2
    *   HTTP::Status

AUTHOR

    Ryo Anazawa (r-anazawa@shochutairen.com)

LICENSE AND COPYRIGHT

    Copyright (c) 2011 Ryo Anazawa. All rights reserved.

    This module is free software; you can redistribute it and/or
    modify it under the same terms as Perl itself. See L<perlartistic>.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



Changes
Revision history for Perl extension Blosxom::Header.

  - target() is obsolete and will be removed in 0.07
  - requires CGI::Header 0.03
  - Although this module shouldn't depend on a global variable
    $blosxom::header (and also shouldn't be a singleton class),
    Blosxom.pm doesn't exist on CPAN. I don't know how to solve
    this problem :( How can I justify this module on CPAN?

0.06003 Aug 19th, 2012
  - push_p3p_tags() is obsolete and will be removed in 0.07
  - new() become a private method, not the alias of instance() any more
  - Add EXAMPLES to POD

0.06002 Aug 9th, 2012
  - Use HTTP::Headers::Util to parse the Content-Type header

0.06001 Aug 2nd, 2012
  - Corrected typo:
      P3P: policyref="/w3c/p3p.xml", CP="%s" (comma was missing)

0.06000 Jul 31st, 2012
  - charset() parse a quoted string
      $header->set( Content_Type => 'text/plain; charset="utf-8"' );
      my $charset = $header->charset; # UTF-8
  - Removed the following methods/functions:
      * cookie
      * push_cookie
      * p3p
      * push_p3p
      * header_push
      * each_header

0.05012 Jul 29th, 2012
  - Removed obsolete tests

0.05010 Jul 28th, 2012
  - Add new() method which is the alias of instance()
  - p3p() and push_p3p() were renamed to p3p_tags() and push_p3p_tags(),
    respectively. p3p() and push_p3p() methods will be removed in 0.06

0.05009 Jul 26th, 2012
  - Add as_hashref() method
  - each() will return nothing in 0.06
  - Exports header_iter() on demand
  - each_header() is obsolete and will be removed in 0.06.
    Use header_iter() instead
  - Add content_type() as the alias of type()

0.05008 Jul 20th, 2012
  - Add is_empty() and flatten()
  - cookie() and push_cookie() is obsolete and will be removed in 0.06.
    These method was replaced with set_cookie() and get_cookie()

0.05007 Jul 13, 2012
  - Add date() and last_modified()
  - When cookie() or push_cookie() receives hash references,
    these methods create CGI::Cookie objects
  - Requires CGI::Cookie
  - Use 'parent' instead of 'base'

0.05006 Jul 7, 2012
  - add each() method
  - requires HTTP::Date and CGI::Util
  - expires() always converts its value to machine time
  - $Header isn't exported any more. Sorry for incovenience :(

0.05005 Jun 30, 2012
  - Exports push_cookie() and push_p3p() instead of header_push().
    header_push() will be removed in 0.06

0.05004 Jun 21, 2012
  - Added field_names()
  - Exports header_push() on demand
  - charset() became read-only
  - p3p() and cookie() return LISTs
  - Modified a difinition of CLEAR()

0.05003 Jun 16, 2012
  - Exports header_get(), header_set(), header_exists(), header_delete()
    on demand
  - Exports the instance as $Header on demand
  - charset() became consistent with type()
  - Added has_instance()

0.05002 Jun 13, 2012
  - Added is_initialized() to Blosxom::Header
  - Added DIAGNOSTICS to POD

0.05001 Jun 13th, 2012
  - Added Blosxom::Header::Proxy
  - Added _tied() method which returns Blosxom::Header::Proxy object
  - Removed _normalize_field_name()
  - status() carps if a given status code is unknown to HTTP::Status

0.04003 May 25th, 2012
  - status() receives the status code:
      $header->status(304)
    Don't pass a string which contains reason phrases:
      $header->status('304 Not Modified') # OBSOLETE
    status() returns the status code:
      $status = $header->status # 304
  - Requires HTTP::Status
  - A tied hash became read-only by default. When you need to modify it,
    you must pass an additional argument 'rw' to TIEHASH() explicitly:
      tie %header => 'Blosxom::Header', 'rw';

0.04002 May 21th, 2012
  - _normalize_field_name() became a class method internal
  - cookie() and p3p() receive LIST

0.04001 May 10th, 2012
  - Removed Blosxom::Header::Class
  - Fixed compilation error

0.04000 May 9th, 2012
  - Removed push(), new() and has_instance() methods
  - Added Blosxom::Header::Class

0.03005 May 4th, 2012
  - Restricted the instantiation of Blosxom::Header to one object
  - Added instance() and has_instance() methods.
  - new() is deprecated and will be removed in 0.04.
    Use instance() instead.

0.03004 Apr 26th, 2012
  - Wrote Japanese document

0.03003 Apr 25th, 2012
  - added clear() and status() methods
  - implemented tie() interface (EXPERIMENTAL)

0.03002 Apr 20th, 2012
  - push() is deprecated and will be removed in 0.04.
    Use push_cookie() or push_p3p() instead.
    Internally, push() was renamed to _push().
  - When we specify field names, we follow HTTP::Headers' way:
      "To make the life easier for perl users who wants to avoid quoting
      before the => operator, you can use '_' as a replacement for '-'
      in header names."
  - Added the following accessors:
      * attachment
      * charset

lib/Blosxom/Header.pm
package Blosxom::Header;
use 5.008_009;
use strict;
use warnings;
use parent 'CGI::Header';
use overload '%{}' => 'as_hashref', bool => 'boolify', fallback => 1;
use Carp qw/carp croak/;
use Exporter 'import';
use Scalar::Util qw/refaddr/;

our $VERSION = '0.06003';

our @EXPORT_OK = qw(
    header_get    header_set  header_exists
    header_delete header_iter
);

sub header_get    { __PACKAGE__->instance->get( @_ )    }
sub header_set    { __PACKAGE__->instance->set( @_ )    }
sub header_exists { __PACKAGE__->instance->exists( @_ ) }
sub header_delete { __PACKAGE__->instance->delete( @_ ) }
sub header_iter   { __PACKAGE__->instance->each( @_ )   }

our $INSTANCE;

sub new { croak "Private method 'new' called for $_[0]" }

sub is_initialized { ref $blosxom::header eq 'HASH' }

sub has_instance { $INSTANCE }

sub instance {
    my $class = shift;

    return $class    if ref $class;
    return $INSTANCE if defined $INSTANCE;

    if ( ref $blosxom::header eq 'HASH' ) {
        return $INSTANCE = $class->SUPER::new( $blosxom::header );
    }

    croak "$class hasn't been initialized yet";
}

sub boolify { 1 }

my %header;

sub as_hashref {
    my $self = shift;
    $header{ refaddr $self } ||= do {
        tie my %header, 'CGI::Header', $self->header;
        \%header;
    };
}

sub DESTROY {
    my $self = shift;
    delete $header{ refaddr $self };
    $self->SUPER::DESTROY;
}

sub get {
    my ( $self, @fields ) = @_;
    my @values = map { $self->FETCH($_) } @fields;
    wantarray ? @values : $values[-1];
}

sub set {
    my ( $self, @headers ) = @_;

    if ( @headers % 2 == 0 ) {
        while ( my ($field, $value) = splice @headers, 0, 2 ) {
            $self->STORE( $field => $value );
        }
    }
    else {
        croak 'Odd number of elements passed to set()';
    }

    return;
}

sub delete {
    my ( $self, @fields ) = @_;

    if ( wantarray ) {
        return map { $self->DELETE($_) } @fields;
    }
    elsif ( defined wantarray ) {
        my $value = @fields && $self->DELETE( pop @fields );
        $self->DELETE( $_ ) for @fields;
        return $value;
    }
    else {
        $self->DELETE( $_ ) for @fields;
    }

    return;
}

# this method is obsolete and will be removed in 0.07
sub push_p3p_tags {
    my ( $self, @tags ) = @_;

    unless ( @tags ) {
        Carp::carp 'Useless use of push_p3p_tags() with no values';
        return;
    }

    if ( my $tags = $self->header->{-p3p} ) {
        return push @{ $tags }, @tags if ref $tags eq 'ARRAY';
        unshift @tags, $tags;
    }

    $self->header->{-p3p} = @tags > 1 ? \@tags : $tags[0];

    scalar @tags;
}

sub charset {
    my $self = shift;

    require HTTP::Headers::Util;

    my %param = do {
        my $type = $self->FETCH( 'Content-Type' );
        my ( $params ) = HTTP::Headers::Util::split_header_words( $type );
        return unless $params;
        splice @{ $params }, 0, 2;
        @{ $params };
    };

    if ( my $charset = $param{charset} ) {
        $charset =~ s/^\s+//;
        $charset =~ s/\s+$//;
        return uc $charset;
    }

    return;
}

sub content_type {
    my $self = shift;

    if ( @_ ) {
        my $content_type = shift;
        $self->STORE( 'Content-Type' => $content_type );
        return;
    }

    my ( $media_type, $rest ) = do {
        my $content_type = $self->FETCH( 'Content-Type' );
        return q{} unless defined $content_type;
        split /;\s*/, $content_type, 2;
    };

    $media_type =~ s/\s+//g;
    $media_type = lc $media_type;

    wantarray ? ($media_type, $rest) : $media_type;
}

BEGIN { *type = \&content_type }

sub date          { shift->_date_header( 'Date',          @_ ) }
sub last_modified { shift->_date_header( 'Last-Modified', @_ ) }

sub expires {
    my $self = shift;

    if ( @_ ) {
        $self->SUPER::expires( shift );
        return;
    }

    $self->_date_header( 'Expires' );
}

sub _date_header {
    my ( $self, $field, $time ) = @_;

    require HTTP::Date;

    if ( defined $time ) {
        $self->STORE( $field => HTTP::Date::time2str($time) );
    }
    elsif ( my $date = $self->FETCH($field) ) {
        return HTTP::Date::str2time( $date );
    }

    return;
}

sub set_cookie {
    my ( $self, $name, $value ) = @_;

    require CGI::Cookie;

    my $cookies = $self->FETCH( 'Set-Cookie' );

    unless ( ref $cookies eq 'ARRAY' ) {
        $cookies = $cookies ? [ $cookies ] : [];
        $self->STORE( 'Set-Cookie' => $cookies );
    }

    my $new_cookie = CGI::Cookie->new(do {
        my %args = ref $value eq 'HASH' ? %{ $value } : ( value => $value );
        $args{name} = $name;
        \%args;
    });

    for my $cookie ( @{$cookies} ) {
        next unless ref $cookie eq 'CGI::Cookie';
        next unless $cookie->name eq $name;
        $cookie = $new_cookie;
        undef $new_cookie;
        last;
    }

    push @{ $cookies }, $new_cookie if $new_cookie;

    return;
}

sub get_cookie {
    my ( $self, $name ) = @_;

    my @cookies = do {
        my $cookies = $self->FETCH( 'Set-Cookie' );
        return unless $cookies;
        ref $cookies eq 'ARRAY' ? @{ $cookies } : $cookies;
    };

    my @values;
    for my $cookie ( @cookies ) {
        next unless ref $cookie eq 'CGI::Cookie';
        next unless $cookie->name eq $name;
        push @values, $cookie;
    }

    wantarray ? @values : $values[0];
}

sub status {
    my $self = shift;

    require HTTP::Status;

    if ( @_ ) {
        my $code = shift;
        my $message = HTTP::Status::status_message( $code );
        return $self->STORE( Status => "$code $message" ) if $message;
        carp "Unknown status code '$code' passed to status()";
    }
    elsif ( my $status = $self->FETCH('Status') ) {
        return substr( $status, 0, 3 );
    }
    #else {
    #    return 200;
    #}

    return;
}

# this method is obsolete and will be remove in 0.07
sub target {
    my $self = shift;
    return $self->STORE( 'Window-Target' => shift ) if @_;
    $self->FETCH( 'Window-Target' );
}

1;

__END__

=head1 NAME

Blosxom::Header - Object representing CGI response headers

=head1 SYNOPSIS

  # object-oriented interface
  use Blosxom::Header;

  my $header = Blosxom::Header->instance;

  $header->set( 'Content-Length' => 12345 );
  my $value   = $header->get( 'Status' );
  my $deleted = $header->delete( 'Content-Disposition' );
  my $bool    = $header->exists( 'ETag' );

  # as a hash reference
  $header->{'Content-Length'} = 12345;
  my $value   = $header->{'Status'};
  my $deleted = delete $header->{'Content-Disposition'};
  my $bool    = exists $header->{'ETag'};

  # procedural interface
  use Blosxom::Header qw(
      header_get    header_set
      header_exists header_delete
  );

  header_set( 'Content-Length' => 12345 );
  my $value   = header_get( 'Status' );
  my $deleted = header_delete( 'Content-Disposition' );
  my $bool    = header_exists( 'ETag' );

=head1 DESCRIPTION

This module provides Blosxom plugin developers
with an interface to handle L<CGI> response headers.
This class represents a global variable C<$blosxom::header>,
and so it can have only one instance.
Since the variable is a reference to a hash, each header field is
restricted to appear only once except the C<Set-Cookie> header.
In other words, the instance behaves like a hash rather than an array.

=head2 BACKGROUND

Blosxom, an weblog application, globalizes C<$header> which is a hash reference.
This application passes C<$header> to C<CGI::header()> to generate
CGI response headers.

Blosxom plugins may modify C<$header> directly because the variable is
global.
The problem is that there is no agreement with how to normalize C<keys>
of C<$header>.
If we normalized them, plugins would be compatible with each other.

In addition, C<CGI::header()> doesn't behave intuitively.
It's difficult for us to predict what the subroutine will return
unless we execute it.
Although CGI's pod tells us how it will work,
it's painful for lazy people to reread the document
whenever they manipulate CGI response headers.
It lowers their productivity.
While it's easy to replace CGI.pm with other modules,
this module extends Blosxom without rewriting C<blosxom.cgi>.

=head2 CLASS METHODS

=over 4

=item $header = Blosxom::Header->instance

Returns a current Blosxom::Header object instance or create a new one.

C<new()> isn't an alias, any more. C<new()> become a private method
since 0.06003.

=item $header = Blosxom::Header->has_instance

Returns a reference to any existing instance or C<undef> if none is defined.

=item $bool = Blosxom::Header->is_initialized

Returns a Boolean value telling whether C<$blosxom::header> is initialized or
not. Blosxom initializes the variable just before C<blosxom::generate()> is
called. If C<$bool> was false, C<instance()> would throw an exception.

Internally, this method is a shortcut for

  $bool = ref $blosxom::header eq 'HASH';

=back

=head2 INSTANCE METHODS

=over 4

=item $value = $header->get( $field )

=item @values = $header->get( @fields )

Returns the value of one or more header fields.
Accepts a list of field names case-insensitive.
You can use underscores as a replacement for dashes in header names.

  # field names are case-insensitive
  $header->get( 'Content-Length' );
  $header->get( 'content_length' );

=item $header->set( $field => $value )

=item $header->set( $f1 => $v1, $f2 => $v2, ... )

Sets the value of one or more header fields.
Accepts a list of named arguments.

The $value argument must be a plain string, except for when the Set-Cookie
header is specified.
In an exceptional case, $value may be a reference to an array.

  use CGI::Cookie;

  my $cookie1 = CGI::Cookie->new( -name => 'foo' );
  my $cookie2 = CGI::Cookie->new( -name => 'bar' );

  $header->set( Set_Cookie => [ $cookie1, $cookie2 ] );

=item $bool = $header->exists( $field )

Returns a Boolean value telling whether the specified HTTP header exists.

  if ( $header->exists('ETag') ) {
      ....
  }

=item @deleted = $header->delete( @fields )

Deletes the specified fields from HTTP headers.
Returns values of deleted fields.

  $header->delete( qw/Content-Type Content-Length Content-Disposition/ );

=item $header->clear

This will remove all header fields.

  $header->clear;

Internally, this method is a shortcut for

  %{ $blosxom::header } = ( -type => q{} );

=item @fields = $header->field_names

Returns the list of distinct names for the fields present in the header.
The field names have case as returned by C<CGI::header()>.

  my @fields = $header->field_names;
  # => ( 'Set-Cookie', 'Content-length', 'Content-Type' )

In scalar context return the number of distinct field names.

=item $header->each( \&callback )

Apply a subroutine to each header field in turn.
The callback routine is called with two parameters;
the name of the field and a value.
Any return values of the callback routine are ignored.

  $header->each(sub {
      my ( $field, $value ) = @_;
      ...
  });

=item $bool = $header->is_empty

Returns a Boolean value telling whether C<< $header->field_names >>
returns a null array or not.

  $header->clear;

  if ( $header->is_empty ) { # true
      ...
  }

=item @headers = $header->flatten

Returns pairs of fields and values.

  my @headers = $header->flatten;
  # => ( 'Status', '304 Not Modified', 'Content-Type', 'text/plain' )

=item $hashref = $header->as_hashref

Returns a reference to hash which represents header fields.
You can manipulate header fields using the hash syntax.

  $header->as_hashref->{Foo} = 'bar';
  my $value   = $header->as_hashref->{Foo};
  my $deleted = delete $header->as_hashref->{Foo};
  my $bool    = exists $header->as_hashref->{Foo};

Since the hash dereference operator of C<$header> is L<overload>ed
with C<as_hashref()>,
you can omit calling C<as_hashref()> method from the above operations:

  $header->{Foo} = 'bar';
  my $value   = $header->{Foo};
  my $deleted = delete $header->{Foo};
  my $bool    = exists $header->{Foo};

NOTE: You can't iterate over C<$header> using C<CORE::each()>, C<CORE::keys()>
or C<CORE::values()>. Use C<< $header->field_names >> or C<< $header->each >>
instead.

  # not supported
  keys %{ $header };
  values %{ $header };
  each %{ $header };

=item $clone = $header->clone

Returns a copy of this Blosxom::Header object.

=back

=head3 HANDLING COOKIES

=over 4

=item $header->set_cookie( $name => $value )

=item $header->set_cookie( $name => { value => $value, ... } )

Overwrites existent cookie.

  $header->set_cookie( ID => 123456 );

  $header->set_cookie(
      ID => {
          value   => '123456',
          path    => '/',
          domain  => '.example.com',
          expires => '+3M',
      }
  );

=item $cookie = $header->get_cookie( $name )

Returns a L<CGI::Cookie> object whose C<name()> is stringwise equal to C<$name>.

  my $id = $header->get_cookie( 'ID' ); # CGI::Cookie object

  my $value   = $id->value;   # '123456'
  my $path    = $id->path;    # '/'
  my $domain  = $id->domain;  # '.example.com'
  my $expires = $id->expires; # 'Thu, 25 Apr 1999 00:40:33 GMT'

=back

=head3 DATE HEADERS

These methods always convert their value to system time
(seconds since Jan 1, 1970).

=over 4

=item $mtime = $header->date

=item $header->date( $mtime )

This header represents the date and time at which the message
was originated. This method expects machine time when the header
value is set.

  $header->date( time ); # set current date

=item $mtime = $header->expires

=item $header->expires( $mtime )

The Expires header gives the date and time after which the entity should be
considered stale.
You can specify an absolute or relative expiration interval.
The following forms are all valid for this field:

  $header->expires( '+30s' ); # 30 seconds from now
  $header->expires( '+10m' ); # ten minutes from now
  $header->expires( '+1h'  ); # one hour from now
  $header->expires( '-1d'  ); # yesterday
  $header->expires( 'now'  ); # immediately
  $header->expires( '+3M'  ); # in three months
  $header->expires( '+10y' ); # in ten years time

  # at the indicated time & date
  $header->expires( 'Thu, 25 Apr 1999 00:40:33 GMT' );

  # another representation of 'now'
  $header->expires( time );

=item $mtime = $header->last_modified

=item $header->last_modified( $mtime )

This header indicates the date and time at which the resource was
last modified. This method expects machine time when the header value is set.

  # check if document is more than 1 hour old
  if ( my $last_modified = $header->last_modified ) {
      if ( $last_modified < time - 60 * 60 ) {
          ...
      }
  }

=back

=head3 CONVENIENCE METHODS

The following methods were named after parameters recognized by
C<CGI::header()>.
These can both be used to read and to set the value of a header.
The value is set if you pass an argument to the method.
If the given header wasn't defined then C<undef> would be returned.

=over 4

=item $filename = $header->attachment

=item $header->attachment( $filename )

Can be used to turn the page into an attachment.
Represents suggested name for the saved file.

  $header->attachment( 'genome.jpg' );
  my $attachment = $header->attachment; # genome.jpg

  my $disposition = $header->get( 'Content-Disposition' );
  # => 'attachment; filename="genome.jpg"'

=item $charset = $header->charset

Returns the upper-cased character set specified in the Content-Type header.

  $header->content_type( 'text/plain; charset=utf-8' );
  my $charset = $header->charset; # UTF-8

This method doesn't receive any arguments.

  # wrong
  $header->charset( 'euc-jp' );

=item $media_type = $header->content_type

=item ( $media_type, $rest ) = $header->content_type

=item $header->content_type( 'text/html; charset=ISO-8859-1' )

Represents the Content-Type header which indicates the media type of
the message content. C<type()> is an alias.

  $header->content_type( 'text/plain; charset=utf-8' );

The value returned will be converted to lower case, and potential parameters
will be chopped off and returned as a separate value if in an array context.

  my $type = $header->content_type; # 'text/plain'
  my @type = $header->content_type; # ( 'text/plain', 'charset=utf-8' )

If there is no such header field, then the empty string is returned.
This makes it safe to do the following:

  if ( $header->content_type eq 'text/html' ) {
      ...
  }

=item $header->nph

If set to a true value,
will issue the correct headers to work with
a NPH (no-parse-header) script:

  $header->nph( 1 );

=item @tags = $header->p3p_tags

=item $header->p3p_tags( @tags )

Represents the P3P tags.
The parameter can be an array or a space-delimited string.

  $header->p3p_tags( qw/CAO DSP LAW CURa/ );
  $header->p3p_tags( 'CAO DSP LAW CURa' );

  my @tags = $header->p3p_tags; # ( 'CAO', 'DSP', 'LAW', 'CURa' )

In this case, the outgoing header will be formatted as:

  P3P: policyref="/w3c/p3p.xml", CP="CAO DSP LAW CURa"

=item $header->push_p3p_tags( @tags )

This method is obsolete and will be removed in 0.07.

Adds P3P tags to the P3P header.
Accepts a list of P3P tags.

  # get P3P tags
  my @tags = $header->p3p_tags; # ( 'CAO', 'DSP', 'LAW' )

  # add P3P tags
  $header->push_p3p_tags( 'CURa' );

  @tags = $header->p3p_tags; # ( 'CAO', 'DSP', 'LAW', 'CURa' )

=item $code = $header->status

=item $header->status( $code )

Represents HTTP status code.

  $header->status( 304 );
  my $code = $header->status; # 304

cf.

  $header->set( Status => '304 Not Modified' );
  my $status = $header->get( 'Status' ); # 304 Not Modified

=item $header->target

Represents the Window-Target header.

  $header->target( 'ResultsWindow' );
  my $target = $header->target; # ResultsWindow

cf.

  $header->set( Window_Target => 'ResultsWindow' );
  $target = $header->get( 'Window-Target' ); # ResultsWindow

=back

=head2 FUNCTIONS

The following functions are exported on demand.

=over 4

=item @values = header_get( @fields )

A shortcut for

  @values = Blosxom::Header->instance->get( @fields );

=item header_set( $field => $value )

=item header_set( $f1 => $v2, $f2 => $v2, ... )

A shorcut for

  Blosxom::Header->instance->set( $field => $value );
  Blosxom::Header->instance->set( $f1 => $v1, $f2 => $v2, ... );

=item $bool = header_exists( $field )

A shortcut for

  $bool = Blosxom::Header->instance->exists( $field );

=item @deleted = header_delete( @fields )

A shorcut for

  @deleted = Blosxom::Header->instance->delete( @fields );

=item header_iter( \&callback )

A shortcut for 

    Blosxom::Header->instance->each( \&callback );

=back

=head1 EXAMPLES

The following script is a Blosxom plugin which just adds the Content-Length
header to CGI response headers.

  package content_length;
  use strict;
  use warnings;
  use Blosxom::Header qw/header_set/;

  sub start { !$blosxom::static_entries }

  sub last { header_set( 'Content-Length' => length $blosxom::output ) }

  1;

=head1 LIMITATIONS

Each header field is restricted to appear only once,
except for the Set-Cookie header.
That's why C<$header> can't C<push()> header fields unlike L<HTTP::Headers>
objects. In other words, C<CGI::header()> behaves like a hash rather than
an array.

=head2 THE P3P HEADER

Since C<CGI::header()> restricts where the policy-reference file is located,
you can't modify the location (C</w3c/p3p.xml>).
The subroutine outputs the P3P header in the following format:

  P3P: policyref="/w3c/p3p.xml", CP="%s"

therefore the following code doesn't work as you expect:

  # wrong
  $header->set( P3P => q{policyref="/path/to/p3p.xml"} );

You're allowed to set or add P3P tags by using C<< $header->p3p_tags >>
or C<< $header->push_p3p_tags >>.

=head2 THE DATE HEADER

C<CGI::header()> fixes the Date header when any of
C<< $header->get( 'Set-Cookie' ) >>,
C<< $header->expires >> or C<< $header->nph >> returns true. 
When the Date header is fixed, you can't modify the value:

  $header->set_cookie( ID => 123456 );
  # => CGI::header() fixes the Date header

  my $bool = $header->date == CORE::time(); # true

  # wrong
  $header->date( $time );
  $header->set( Date => $date );

=head1 DIAGNOSTICS

=over 4

=item Blosxom::Header hasn't been initialized yet

You attempted to create a Blosxom::Header object
before C<$blosxom::header> was initialized.
See C<< Blosxom::Header->is_initialized() >>.

=item Unknown status code '%03d' passed to status()

The given status code is unknown to L<HTTP::Status>.

=item The Date header is fixed

You attempted to C<set()> or C<delete()> the Date header
when the Date header was fixed. See L<"LIMITATIONS">.

=item Can't assign to '%s' directly, use accessors instead

You attempted to C<set()> the Expires or P3P header.
You can't assign any values to these headers directly.
Use C<expires()> or C<p3p_tags()> instead.

=back

=head1 DEPENDENCIES

L<Blosxom 2.0.0|http://blosxom.sourceforge.net/> or higher.

=head1 SEE ALSO

L<HTTP::Headers>,
L<Plack::Util>,
L<Class::Singleton>

=over 4

=item D. Robinson and K.Coar, "The Common Gateway Interface (CGI) Version 1.1",
L<RFC 3875|http://tools.ietf.org/html/rfc3875#section-6>, October 2004

=back

=head1 ACKNOWLEDGEMENT

Blosxom was originally written by Rael Dornfest.
L<The Blosxom Development Team|http://sourceforge.net/projects/blosxom/>
succeeded to the maintenance.

=head1 BUGS

There are no known bugs in this module.
Please report problems to ANAZAWA (anazawa@cpan.org).
Patches are welcome.

=head1 AUTHOR

Ryo Anazawa (anazawa@cpan.org)

=head1 LICENSE

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.

=cut

